### PR TITLE
[19.01] Add datatypes and sniffers for tar archives of directory-based MS formats and WIFF/SCAN pairs

### PR DESCRIPTION
**Pull request by chambm**
Adds datatypes and sniffers for Bruker BAF, Agilent/Bruker YEP, Bruker TDF, Agilent MassHunter, and Waters MassLynx "raw tar" files.

### CODE REVIEW

**Comment by bgruening** [comment]
@chambm can you please fix the linting errors?

File: lib/galaxy/datatypes/binary.py [hunk 1]
```diff
@@ -2239,6 +2239,126 @@ def sniff(self, dataset):
...
+
+    def display_peek(self, dataset):
+        try:
+            return dataset.peek
+        except Exception:
+            return "%s (%s)" % (self.get_type(), nice_size(dataset.get_size()))
+
+
+class YepTar(BafTar):
+    """ A tar'd up .d directory containing Agilent/Bruker YEP format data """
```
Comment (bgruening) [inline lib/galaxy/datatypes/binary.py:2285]: What is this .d in the docstring about?

**Review by bgruening — APPROVED** [review]

File: lib/galaxy/datatypes/binary.py (hunk 1 above)
Comment (chambm) [reply lib/galaxy/datatypes/binary.py:2285]: The directories (almost) always end in .d. That's how users are supposed to be able to tell the difference between them and a regular old folder when browsing the filesystem. Too bad that convention is used for half a dozen different MS data formats...

File: lib/galaxy/datatypes/binary.py [hunk 2]
```diff
@@ -2239,6 +2239,126 @@ def sniff(self, dataset):
...
+    """
+    edam_data = "data_2536"  # mass spectrometry data
+    edam_format = "format_3712"  # TODO: add more raw formats to EDAM?
+    file_ext = "brukerbaf.tar"
+
+    def get_signature_file(self):
+        return "analysis.baf"
+
+    def sniff(self, filename):
+        try:
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: This is unnecessary, please remove the try/except around the sniff, that's caught outside. We really don't want to log anything here.

File: lib/galaxy/datatypes/binary.py [hunk 3]
```diff
@@ -2239,6 +2239,126 @@ def sniff(self, dataset):
...
+    >>> WiffTar().sniff(fname)
+    False
+    >>> fname = get_test_fname('test.fast5.tar')
+    >>> WiffTar().sniff(fname)
+    False
+    """
+    file_ext = "wiff.tar"
+
+    def sniff(self, filename):
+        try:
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]: Same here

**Comment by mvdbeek** [comment]
@galaxybot test py3

**Comment by mvdbeek** [comment]
@galaxybot test this

**Comment by nsoranzo** [comment]
@galaxybot test py3

**Comment by mvdbeek** [comment]
screen shot 2019-02-06 at 12 30 06
I think we need to set this for all the py3 variants. @martenson did you update the PR builder plugin recently ? Any idea why that has diverged between the jobs ?

**Comment by mvdbeek** [comment]
@galaxybot test py3

**Comment by mvdbeek** [comment]
Yep, that was the issue ...

File: lib/galaxy/datatypes/binary.py [hunk 4]
```diff
@@ -2239,6 +2239,126 @@ def sniff(self, dataset):
...
+    >>> fname = get_test_fname('brukerbaf.d.tar')
+    >>> BafTar().sniff(fname)
+    True
+    >>> fname = get_test_fname('test.fast5.tar')
+    >>> BafTar().sniff(fname)
+    False
+    """
+    edam_data = "data_2536"  # mass spectrometry data
+    edam_format = "format_3712"  # TODO: add more raw formats to EDAM?
+    file_ext = "brukerbaf.tar"
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]:
```suggestion
    file_ext = "brukerbaf.d.tar"
```

File: lib/galaxy/datatypes/binary.py [hunk 5]
```diff
@@ -2239,6 +2239,126 @@ def sniff(self, dataset):
...
+    def display_peek(self, dataset):
+        try:
+            return dataset.peek
+        except Exception:
+            return "%s (%s)" % (self.get_type(), nice_size(dataset.get_size()))
+
+
+class YepTar(BafTar):
+    """ A tar'd up .d directory containing Agilent/Bruker YEP format data """
+    file_ext = "agilentbrukeryep.tar"
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]:
```suggestion
    file_ext = "agilentbrukeryep.d.tar"
```

File: lib/galaxy/datatypes/binary.py [hunk 6]
```diff
@@ -2239,6 +2239,126 @@ def sniff(self, dataset):
...
+    def get_signature_file(self):
+        return "analysis.yep"
+
+    def get_type(self):
+        return "Agilent/Bruker YEP directory archive"
+
+
+class TdfTar(BafTar):
+    """ A tar'd up .d directory containing Bruker TDF format data """
+    file_ext = "brukertdf.tar"
```
Comment (mvdbeek) [inline lib/galaxy/datatypes/binary.py]:
```suggestion
    file_ext = "brukertdf.d.tar"
```

**Review by mvdbeek — COMMENTED** [review]
To get the tests to pass you need to either use `"<variant>.d.tar"` as file_ext or rename the test data to match the extension

**Comment by chambm** [comment]
When you say "get tests to pass", which tests? Where do I see the outcome?

Nevermind...I see those tests were just added later after the others had passed.

**Comment by nsoranzo** [comment]
@galaxybot test this

**Comment by mvdbeek** [comment]
https://jenkins.galaxyproject.org/job/docker-integration/7764/testReport/junit/test.integration/test_datatype_upload/test_upload_datatype_auto_brukerbaf_d_tar_/ for instance

These are automatically constructed for test files in lib/galaxy/datatypes/test.
We upload them, make sure the dataset sniffed is of the expected type and that it doesn't generate false positive matches.

**Comment by mvdbeek** [comment]
@galaxybot test py3

**Comment by bgruening** [comment]
Thanks all!

**Comment by martenson** [comment]
@mvdbeek 
> did you update the PR builder plugin recently

that could have happened, I update all plugins regularly

**Pull request merged**